File: ios/FiHavenCore/Sources/FiHavenCore/Logic/Payoff.swift
import Foundation

public enum PayoffStrategy: String, Sendable, CaseIterable {
    case none       // pay minimums only
    case snowball   // extra → smallest balance first
    case avalanche  // extra → highest APR first
}

public struct PayoffCardResult: Equatable, Sendable, Identifiable {
    public var id: String
    public var name: String
    public var origBalance: Double
    public var paidOffMonth: Int?
    public var interestPaid: Double
}

public struct PayoffResult: Equatable, Sendable {
    public var months: Int
    public var totalInterest: Double
    public var cards: [PayoffCardResult]
    public var payoffDate: Date
}

/// Month-by-month debt-payoff simulation, ported verbatim from
/// payoff.js `runPayoffSim`. Interest accrues monthly
/// (`regularAPR/100/12`) and is skipped while a card is inside its 0%
/// promo window; freed minimums roll into the extra pool (the snowball
/// rollover). Capped at 360 months.
///
/// Returns nil when no card carries a balance.
public enum Payoff {
    private struct Sim {
        var id: String
        var name: String
        var balance: Double
        var origBalance: Double
        var minPayment: Double
        var apr: Double
        var monthlyRate: Double
        var hasPromo: Bool
        var promoEnd: Date?
        var paidOffMonth: Int?
        var interestPaid: Double
        var housing: Bool
    }

    /// Mortgage / home-equity style loans — PMI & escrow make sims approximate.
    public static func isHousingLoan(_ c: Card) -> Bool {
        guard (c.type ?? "card") == "loan" else { return false }
        let hay = [c.name, c.issuer ?? ""]
            .joined(separator: " ")
            .lowercased()
        let patterns = ["mortgage", "home equity", "heloc", "housing", "home loan", "refinance", "refi"]
        return patterns.contains { hay.contains($0) }
    }

    public static func runPayoffSim(
        cards: [Card],
        strategy: PayoffStrategy,
        extra: Double,
        tz: TimeZone,
        now: Date = Date(),
        includeMortgage: Bool = false
    ) -> PayoffResult? {
        // Prefer live Current Balance when set (matches web/Android payoff.js).
        func debtOf(_ c: Card) -> Double {
            if let cur = c.currentBalance, cur > 0 { return cur }
            return c.balance
        }
        let debtCards = cards.filter { debtOf($0) > 0 && (includeMortgage || !isHousingLoan($0)) }
        guard !debtCards.isEmpty else { return nil }

        var sim = debtCards.map { c in
            let starting = debtOf(c)
            let isLoan = (c.type ?? "card") == "loan"
            return Sim(
                id: c.id,
                name: c.name,
                balance: starting,
                origBalance: starting,
                minPayment: max(c.minPayment, 1),
                apr: c.regularAPR,
                monthlyRate: c.regularAPR / 100 / 12,
                hasPromo: isLoan ? false : c.hasPromo,
                promoEnd: isLoan ? nil : DateLogic.parseDate(c.promoEndDate, tz: tz),
                paidOffMonth: nil,
                interestPaid: 0,
                housing: isHousingLoan(c)
            )
        }

        switch strategy {
        case .snowball: sim.sort { $0.origBalance < $1.origBalance }
        case .avalanche: sim.sort { $0.apr > $1.apr }
        case .none: break
        }

        let cal = DateLogic.calendar(tz: tz)
        let nowComps = cal.dateComponents([.year, .month], from: now)
        let baseYear = nowComps.year ?? 0
        let baseMonth = nowComps.month ?? 1

        var month = 0
        var totalInterest = 0.0
        var extraPool = extra

        while sim.contains(where: { $0.balance > 0.01 }) && month < 360 {
            month += 1
            let targetDate = DateLogic.dateForDay(1, year: baseYear, month: baseMonth + month, cal: cal)

            // Accrue interest (skipped inside a promo window).
            for i in sim.indices {
                guard sim[i].balance > 0.01 else { continue }
                let inPromo = sim[i].hasPromo
                    && (sim[i].promoEnd.map { $0 >= targetDate } ?? false)
                if !inPromo && sim[i].monthlyRate > 0 {
                    let interest = sim[i].balance * sim[i].monthlyRate
                    sim[i].interestPaid += interest
                    totalInterest += interest
                    sim[i].balance += interest
                }
            }

            // Pay each minimum.
            var freedThisMonth = 0.0
            for i in sim.indices {
                guard sim[i].balance > 0.01 else { continue }
                let pay = min(sim[i].balance, sim[i].minPayment)
                sim[i].balance -= pay
                if sim[i].balance < 0.01 {
                    sim[i].balance = 0
                    if sim[i].paidOffMonth == nil {
                        sim[i].paidOffMonth = month
                        freedThisMonth += sim[i].minPayment
                    }
                }
            }

            // Apply the extra pool down the sorted list.
            if strategy != .none && extraPool > 0.01 {
                var remaining = extraPool
                for i in sim.indices {
                    if remaining <= 0.01 { break }
                    guard sim[i].balance > 0.01 else { continue }
                    let pay = min(sim[i].balance, remaining)
                    sim[i].balance -= pay
                    remaining -= pay
                    if sim[i].balance < 0.01 {
                        sim[i].balance = 0
                        if sim[i].paidOffMonth == nil {
                            sim[i].paidOffMonth = month
                            freedThisMonth += sim[i].minPayment
                        }
                    }
                }
            }

            extraPool += freedThisMonth
        }

        return PayoffResult(
            months: month,
            totalInterest: (totalInterest * 100).rounded() / 100,
            cards: sim.map {
                PayoffCardResult(
                    id: $0.id,
                    name: $0.name,
                    origBalance: $0.origBalance,
                    paidOffMonth: $0.paidOffMonth,
                    interestPaid: $0.interestPaid
                )
            },
            payoffDate: DateLogic.dateForDay(1, year: baseYear, month: baseMonth + month, cal: cal)
        )
    }
}

File: android/core/src/main/kotlin/app/fihaven/core/logic/Payoff.kt
package app.fihaven.core.logic

import app.fihaven.core.model.Card
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToLong

enum class PayoffStrategy { NONE, SNOWBALL, AVALANCHE }

data class PayoffCardResult(
    val id: String,
    val name: String,
    val origBalance: Double,
    val paidOffMonth: Int?,
    val interestPaid: Double,
)

data class PayoffResult(
    val months: Int,
    val totalInterest: Double,
    val cards: List<PayoffCardResult>,
    val payoffDate: LocalDate,
)

/// Month-by-month debt-payoff simulation, ported verbatim from
/// payoff.js `runPayoffSim`. Interest accrues monthly and is skipped while
/// a card is inside its 0% promo window; freed minimums roll into the
/// extra pool. Capped at 360 months. Null when no card carries a balance.
object Payoff {
    private class Sim(
        val id: String,
        val name: String,
        var balance: Double,
        val origBalance: Double,
        val minPayment: Double,
        val apr: Double,
        val monthlyRate: Double,
        val hasPromo: Boolean,
        val promoEnd: LocalDate?,
        var paidOffMonth: Int?,
        var interestPaid: Double,
        val housing: Boolean,
    )

    /** Mortgage / home-equity loans — PMI & escrow make sims approximate. */
    fun isHousingLoan(c: Card): Boolean {
        if (c.type != "loan") return false
        val hay = listOfNotNull(c.name, c.issuer).joinToString(" ").lowercase()
        return listOf("mortgage", "home equity", "heloc", "housing", "home loan", "refinance", "refi")
            .any { hay.contains(it) }
    }

    fun runPayoffSim(
        cards: List<Card>,
        strategy: PayoffStrategy,
        extra: Double,
        zone: ZoneId,
        now: Instant = Instant.now(),
        includeMortgage: Boolean = false,
    ): PayoffResult? {
        val debtCards = cards.filter {
            (it.currentBalance ?: it.balance) > 0 && (includeMortgage || !isHousingLoan(it))
        }
        if (debtCards.isEmpty()) return null

        val sim = debtCards.map { c ->
            val startingBalance = c.currentBalance ?: c.balance
            val isLoan = c.type == "loan"
            Sim(
                id = c.id,
                name = c.name,
                balance = startingBalance,
                origBalance = startingBalance,
                minPayment = max(c.minPayment, 1.0),
                apr = c.regularAPR,
                monthlyRate = c.regularAPR / 100.0 / 12.0,
                hasPromo = if (isLoan) false else c.hasPromo,
                promoEnd = if (isLoan) null else DateLogic.parseDate(c.promoEndDate),
                paidOffMonth = null,
                interestPaid = 0.0,
                housing = isHousingLoan(c),
            )
        }.toMutableList()

        when (strategy) {
            PayoffStrategy.SNOWBALL -> sim.sortBy { it.origBalance }
            PayoffStrategy.AVALANCHE -> sim.sortByDescending { it.apr }
            PayoffStrategy.NONE -> {}
        }

        val firstThisMonth = DateLogic.today(zone, now).withDayOfMonth(1)
        var month = 0
        var totalInterest = 0.0
        var extraPool = extra

        while (sim.any { it.balance > 0.01 } && month < 360) {
            month++
            val targetDate = firstThisMonth.plusMonths(month.toLong())

            // Accrue interest (skipped inside a promo window).
            for (c in sim) {
                if (c.balance <= 0.01) continue
                val inPromo = c.hasPromo && c.promoEnd != null && !c.promoEnd.isBefore(targetDate)
                if (!inPromo && c.monthlyRate > 0) {
                    val interest = c.balance * c.monthlyRate
                    c.interestPaid += interest
                    totalInterest += interest
                    c.balance += interest
                }
            }

            // Pay each minimum.
            var freedThisMonth = 0.0
            for (c in sim) {
                if (c.balance <= 0.01) continue
                val pay = min(c.balance, c.minPayment)
                c.balance -= pay
                if (c.balance < 0.01) {
                    c.balance = 0.0
                    if (c.paidOffMonth == null) {
                        c.paidOffMonth = month
                        freedThisMonth += c.minPayment
                    }
                }
            }

            // Apply the extra pool down the sorted list.
            if (strategy != PayoffStrategy.NONE && extraPool > 0.01) {
                var remaining = extraPool
                for (c in sim) {
                    if (remaining <= 0.01) break
                    if (c.balance <= 0.01) continue
                    val pay = min(c.balance, remaining)
                    c.balance -= pay
                    remaining -= pay
                    if (c.balance < 0.01) {
                        c.balance = 0.0
                        if (c.paidOffMonth == null) {
                            c.paidOffMonth = month
                            freedThisMonth += c.minPayment
                        }
                    }
                }
            }

            extraPool += freedThisMonth
        }

        return PayoffResult(
            months = month,
            totalInterest = (totalInterest * 100).roundToLong() / 100.0,
            cards = sim.map {
                PayoffCardResult(it.id, it.name, it.origBalance, it.paidOffMonth, it.interestPaid)
            },
            payoffDate = firstThisMonth.plusMonths(month.toLong()),
        )
    }
}
